HATRA '21 paper corrections (#57)
Roblox/luau: papers/hatra21/hatra21.tex

\label{fig:studio}
\end{figure}

At some point during experience design, the experience creator has a need
which can't be met by the physics engine alone, such as ``The stairs should
light up when a player walks on them'' or ``a firework is set off
every few seconds.'' At this point they will discover the script
editor, seen in Fig.~\ref{fig:studio}(b).

This onboarding experience is different from many initial exposures to
programming, in that by the time the user first opens the script
editor, they have already built much of their creation, and have a
very specific concrete aim. As such, Luau must allow users to perform a

> specific task with a minimum of ``boilerplate'' code.

Reviewer: Hmm, the lack of boilerplate is more about the Roblox API than Luau or Studio.
```suggestion
specific task with as much help as possible from tools.
```

\section{Types}
\subsection{Infallible types}

> Goal: \emph{Provide type information even for ill-typed or syntactically invalid programs.}

Reviewer: Nit: capitalization
```suggestion
Goal: \emph{provide type information even for ill-typed or syntactically invalid programs.}
```

where $M \rightarrow M'$ implies $\squnder{M} \rightarrow \squnder{M'}$, and
for any value $V$ we have $\squnder{V} \rightarrow V$, then show:
\begin{itemize}

> \item \emph{Progress}: if ${} \vdash M \Rightarrow N : T$, then $N \rightarrow N'$ or $N$ is a value or $N$ has a flagged subterm.

Reviewer: Nit: in mathematical English, I prefer using "if X then either Y or Z" since "if X then Y or Z" is ambiguous, it could mean "(if X then Y) or Z"."
```suggestion
\item \emph{Progress}: if ${} \vdash M \Rightarrow N : T$, then either $N \rightarrow N'$ or $N$ is a value or $N$ has a flagged subterm.
```

  to find all the possible types a property might be updated with?

\item The natural formulation of function types in a nonstrict setting

>   is: if $f: T \rightarrow U$ and $f(V) \rightarrow^* W$

Reviewer: I remembered the citation I was going to put here!
```suggestion
  is that of~\cite{SuccessTyping}: if $f: T \rightarrow U$ and $f(V) \rightarrow^* W$
```

to be able to infer types for symbols without any annotations being present, to the best of
its ability. This precludes syntactical rules such as Rust's requirement that all function
parameters be annotated with their type~\cite[Ch. 3.3]{RustBook}.

This requirement presents challenges for the type inference algorithm, because Luau may not have
enough information to determine the type of a given program. In non-strict mode in particular,
which sees use in existing codebases, we cannot rely on the presence of type annotations. We also
cannot require that the user provide them if Luau cannot deduce the type of a symbol. In cases
such as this, we must admit defeat and assume that the code is correct, to fulfill non-strict
mode's goal of ``no false positives''. We do this by saying that the result of the operation is
$\mathsf{any}$, a type that can be converted to and from any other type freely.

In strict mode, Luau is not so limited, and in pursuit of the strict-mode goal of ``no false
negatives'', we may surface errors to the user indicating that the type inference system requires
more information, in the form of annotations, in order to type-check a piece of code. This code,

> for example, requires a type annotation in order for Luau to determine the return type of the function:

Reviewer: For people not familiar with Luau, it's worth spelling out why.
```suggestion
for example, requires a type annotation in order for Luau to determine the return type of the function (since Luau does not know if ``+'' refers to built-in addition on numbers, or a user-defined method):
```

\lstset{language=[5.1]Lua}
\begin{lstlisting}
function f(a, b)
  return a + b
end
\end{lstlisting}

Some questions raised by type inference:
\begin{itemize}

\item How many cases in strict mode cannot be inferred by the type inference system? Minimizing
  this kind of error is desirable, to make the type system as unobtrusive as possible.

> 

Reviewer: ```suggestion
\item Can something like the Rust traits system~\cite{RustBook} or Haskell classes~\cite{TypeClasses} be used to provide types for overloaded operators, without hopelessly confusing learners?

\item Can type inference be used to infer the same types in strict and nonstrict mode, to ease migrating between modes, with the only difference being error reporting?
```
